### ISSUE TITLE
How can we differentiate error classes types?

### ISSUE THREAD

**Issue by jrmdayn** [ehmicky/modern-errors#17]
### Guidelines

- [X] Please search other issues to make sure this bug has not already been reported.

### Describe the bug

Hello and thanks for this very cool lib!!

Following the README example:
```ts
import ModernError from 'modern-errors'

export const BaseError = ModernError.subclass('BaseError')

export const UnknownError = BaseError.subclass('UnknownError')
export const InputError = BaseError.subclass('InputError')
export const AuthError = BaseError.subclass('AuthError')
export const DatabaseError = BaseError.subclass('DatabaseError')
```

All error classes have the same (TS) type: `ErrorSubclassCore<[], {}, CustomClass>`, which means that it is impossible to differentiate different errors based on types. For example, the following example compiles:
```ts
declare const unknownError: typeof UnknownError
const inputError: typeof InputError = unknownError // no TS error!!
```

(Also the fact that I have to use `typeof` is a bit annoying, maybe you have a better solution?)

The reason I am asking this is because I am using [fp-ts Either](https://gcanti.github.io/fp-ts/modules/Either.ts.html) in my app and would like to use `modern-errors` for the `E` param, like so:
```ts
type MyType = Either<NotFoundError | UnknownError, number>
```
and be sure that only `NotFoundError` or `UnknownError` are tolerated by the compiler, but it is not the case at the moment..

### Steps to reproduce

playground [here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAWQgEwKZQHYFEpWnAMzxDgHIQV0MBadPKAZ1IChnUAPSWOAYwgwbwAQgEMGqHPTgBeRJUyToAOgYBXAEY8ANmIYAKUqPGKopAJSsOXeHwHwAqhgDWGCAHdsufLKMSvUFQ1tXQNHF3dPenM2TmgbfkE4AEkMMFUYExk4XxNAzR0GfVIUtIz-aKs43gT4AEF0gAtMnzE-ejzgwoN6mCbyi0ruW0SAEREYEXVW5uzp-w6CorGJqeN+1jRgqFRquzhVZ1cPEwAuOBgATzBUCAI4MKPI6BjrXcTgVPTT86ubu5Kvv4sgdwscgQB6cFwVxwAAqAGU4HRoABCFFAA)

### Environment

```
  System:
    OS: macOS 11.3.1
    CPU: (8) arm64 Apple M1
    Memory: 277.31 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.15.1 - ~/.nvm/versions/node/v16.15.1/bin/node
    Yarn: 1.22.5 - ~/.yarn/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v16.15.1/bin/npm
  Browsers:
    Chrome: 109.0.5414.119
    Safari: 14.1
  npmPackages:
    modern-errors: 5.5.1 => 5.5.1 
```

### Pull request (optional)

- [ ] I can submit a pull request.

**Comment by ehmicky** [comment]
Hi @jrmdayn,

Thanks a lot for reaching out.

TypeScript is based [on structural typing](https://imfaber.me/typescript-structural-typing/), as opposed to nominal typing. The error classes above are identical from a functionality standpoint, which results in identical types.

This is not specific to this library: TypeScript itself types native error classes like this. For example, you would get the same issue when using `RangeError` and `URIError`.

A common solution to this problem is to add a tag. For example, the error class' type could have an additional attribute to differentiate it from other classes. For instance, `error.name` could be the a literal string like `"InputError" as const` instead of being a generic `string`. This is actually what this library used to do in previous releases. However, this breaks inheritance: if you create a subclass of `InputError`, TypeScript will not consider it a child type anymore because of the mismatch on that tag attribute.

That being said, if you do not use inheritance, you can circumvent this by using [internal `props`](https://github.com/ehmicky/modern-errors#internal-error-properties).

```ts
import ModernError from 'modern-errors'

export const BaseError = ModernError.subclass('BaseError')

export const UnknownError = BaseError.subclass('UnknownError', { props: { _tag: 'UnknownError' as const } })
export const InputError = BaseError.subclass('InputError', { props: { _tag: 'InputError' as const } })
export const AuthError = BaseError.subclass('AuthError', { props: { _tag: 'AuthError' as const } })
export const DatabaseError = BaseError.subclass('DatabaseError', { props: { _tag: 'DatabaseError' as const } })
```

Would this solution work in your case?

**Comment by jrmdayn** [comment]
Thanks for your response. I understand the solution you offer. I tried it on [TS playground ](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAWQgEwKZQHYFEpWnAMzxDgHIQV0MBadPKAZ1IChnUAPSWOAYwgwbwAQgEMGqHPTgBeRJUyToAOgYBXAEY8ANmIYAKUqPGKopAJSsOXeHwHwAqhgDWGCAHdsufLKMSvUFQ1tXQNHF3dPelIAGjgAbzgwPDAGAC54uAB9GBEAc3TSMNcPE1I4MV5+QTgAX1qLK2gbKvgASQwwVRgTGThfE0DNHQZ9UnbO7v8YjKSIFPSE7LyC8a7S8oZKu1r6tk4mreqAQS6ACx6fMT96QeCRgxOYc6nYhNn5jKX8skfnqI3DvA6jUGvtuLZqgAREQ5dRXC59eH+W7DUbQ2FIqKvRLJNKfHLfUjokRw4xTAEQoG7ZgwACeYFQcCwvTpDIgBDgRQiPQAPnBWah2XBVpN6KxKXBVL0MKg3JznMVItADOY9tZAXBUOlmbJVEA)but I run into a type issue so not sure that it would solve my use case:

```
type E = typeof UnknownError | typeof InputError

const u = new UnknownError('')
export const e: E = u // TS error
```

From a DX point a vue, the `type E` is not super friendly: 
```
type E = ErrorSubclassCore<[], {
    _tag: "UnknownError";
}, CustomClass> | ErrorSubclassCore<[], {
    _tag: "InputError";
}, CustomClass>
```
hard to read if I have more than 2 errors IMO.

**Comment by ehmicky** [comment]
The following should work:

```ts
type E = InstanceType<typeof UnknownError> | InstanceType<typeof InputError>
```

`typeof InputError` is the type of the error class, not the error instance.

The reason this might be surprising and counter-intuitive is because classes in TypeScript work differently. For example, `URIError` is the error instance type, while `typeof URIError` is the error class type. This behavior only happens when using classes:

```ts
class ErrorSubclassCore {
  constructor(...)
}
```

Unfortunately, using classes in TypeScript has the following limitation: the instance type cannot use conditionals. On the other hand, classes can be emulated with interfaces, which do not have this limitation:

```ts
interface ErrorSubclassCore {
  new(...): condition ? valueOne : valueTwo
}
```

Without getting into the details, there are some additional limitations of using classes (as opposed to interfaces) when it comes to using generics for the constructor, and also in terms of typing the `prototype` (which is important for `modern-errors` since it is used for `instanceof` narrowing).

In a nutshell, the error class is typed using an `interface` instead of a `class` to use those additional type features. However, it has the following adverse consequence in terms of developer experience: `InstanceType<typeof ErrorClass>` must be used instead of `ErrorClass` to get the instance type. I am not too happy about this, but I don't quite think there is another way to do this with TypeScript unfortunately :-/

Thanks for pointing out at this issue though. :pray:  I have added some documentation around it in 73ffaf647d2cfb72b036f808b2d5e2464b86cad3, which I hope will help other users.
